app/src/main/java/com/jshvarts/todoapp/arch/MviViewModel.kt

```diff
@@ -11,18 +16,28 @@ import androidx.lifecycle.ViewModel
  * ```
  */
 abstract class MviViewModel<A : UiAction, S : UiState, E : UiEffect>(
-  private val savedStateHandle: SavedStateHandle? = null
-) : ViewModel(), ActionDispatcher<A> {
-  abstract val initialState: S
-  abstract val savedStateHandleKey: String?
-  abstract val actionHandler: (A) -> Unit
-
-  private val isStateInBundle: Boolean
-    get() = if (savedStateHandleKey != null && savedStateHandle != null) {
-      initialState != savedStateHandle[savedStateHandleKey!!]
-    } else false
-
-  override fun dispatchAction(action: A) {
-    return if (isStateInBundle) Unit else actionHandler.invoke(action)
-  }
-}
\ No newline at end of file
+    private val savedStateHandle: SavedStateHandle? = null
+) : ViewModel() {
+    protected abstract val initialState: S
+
+    private val _uiEffect = MutableSharedFlow<E>()
+    val uiEffect: SharedFlow<E> = _uiEffect.asSharedFlow()
+
+    abstract val uiState: StateFlow<S>
```

Comment:
Same, but only the interface. Derived classes will handle the actual implementation.

Reply:
> Same, but only the interface. Derived classes will handle the actual implementation.

👍